src/libmw/src/mmr/Segment.cpp

```diff
@@ -0,0 +1,147 @@
+    for (const Index& idx : hash_indices) {
+        segment.hashes.push_back(mmr.GetHash(idx));
+    }
+
+    // Determine the lowest peak that can be calculated using the hashes we've already provided
+    auto peak = *std::find_if(
+        peak_indices.begin(), peak_indices.end(),
+        [&last_leaf_idx](const Index& peak_idx) { return peak_idx > last_leaf_idx.GetNodeIndex(); });
+
+    // Bag the next lower peak (if there is one), so the root can still be calculated
+    segment.lower_peak = BagNextLowerPeak(mmr, peak_indices, peak, num_nodes);
+
+    return segment;
+}
+
+std::set<Index> SegmentFactory::CalcHashIndices(
+    const ILeafSet& leafset,
+    const std::vector<Index>& peak_indices,
+    const mmr::LeafIndex& first_leaf_idx,
+    const mmr::LeafIndex& last_leaf_idx)
+{
+    std::set<Index> proof_indices;
+
+    // 1. Add peaks of mountains to the left of first index
+    boost::optional<Index> prev_peak = boost::make_optional(false, Index());
+    for (const Index& peak_idx : peak_indices) {
+        if (peak_idx < first_leaf_idx.GetNodeIndex()) {
+            proof_indices.insert(peak_idx);
+            prev_peak = peak_idx;
+        } else {
+            break;
+        }
+    }
+
+    // 2. Add indices needed to reach left edge of mountain
+    auto on_mountain_left_edge = [prev_peak](const Index& idx) -> bool {
+        const uint64_t adjustment = !!prev_peak ? prev_peak->GetPosition() + 1 : 0;
+        return ((idx.GetPosition() + 2) - adjustment) == (uint64_t(2) << idx.GetHeight());
+    };
+
+    Index idx = first_leaf_idx.GetNodeIndex();
+    while (!on_mountain_left_edge(idx)) {
+        Index sibling_idx = idx.GetSibling();
+        if (sibling_idx < idx) {
+            proof_indices.insert(sibling_idx);
+            idx = Index(idx.GetPosition() + 1, idx.GetHeight() + 1);
+        } else {
+            idx = Index(sibling_idx.GetPosition() + 1, sibling_idx.GetHeight() + 1);
+        }
+    }
+
+    // 3. Add all pruned parents after first leaf and before last leaf
+    BitSet pruned_parents = MMRUtil::CalcPrunedParents(leafset.ToBitSet());
+    for (uint64_t pos = first_leaf_idx.GetPosition(); pos < last_leaf_idx.GetPosition(); pos++) {
+        if (pruned_parents.test(pos)) {
+            proof_indices.insert(Index::At(pos));
+        }
+    }
+
+    // 4. Add indices needed to reach right edge of mountain containing the last leaf
+    auto peak_iter = std::find_if(
+        peak_indices.begin(), peak_indices.end(),
+        [&last_leaf_idx](const Index& peak_idx) { return peak_idx > last_leaf_idx.GetNodeIndex(); });
+    assert(peak_iter != peak_indices.end());
+
+    Index peak = *peak_iter;
+    auto on_mountain_right_edge = [prev_peak, peak](const Index& idx) -> bool {
+        const uint64_t adjustment = prev_peak
+                                        .map([](const Index& i) { return i.GetPosition() + 1; })
+                                        .value_or(0);
```

Comment:
I think the code to calculate this should match that in `on_mountain_left_edge`.